### PR TITLE
Add file-level docstring telling what functions do

### PR DESCRIPTION
**Pull request by Asalle**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 1]
```diff
@@ -1,3 +1,12 @@
+"""Functions that extract doc comments.
+
+The only public function extract_documentation_with_markers() is used by
+DocBaseClass.extract(). All it does is going over the code line by line
+and calling _extract_doc_comment_from_line(), which calls
+_extract_doc_comment(), which delegates extracting docs to three more
+specific functions: _extract_doc_comment_simple(),
+_extract_doc_comment_continuous() and _extract_doc_comment_standard().
+"""
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Hey although this is partially correct, I think it misses the point of the functions and how they integrate together.

How about something like this:

```
"""
Functions that extract documentation comments independently of any langage or documentation style.

Each of these functions is built upon one another, and at the last, exposes a single function `extract_documentation_with_markers` which is used by `DocBaseClass`, to extract documentation.
"""
```

Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Just an idea, you can change it however you feel like. Also, make sure you adhere to 79 char line limit though

Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Hi!

Thanks for the review. I applied your suggestions.

**Comment by SanketDG** [comment]
@Asalle don't see the changes :thinking:

**Comment by Asalle** [comment]
So extremely sorry, I forgot to ass when amending D:

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 2]
```diff
@@ -1,3 +1,10 @@
...
+Documentation comments extraction, independently of language or docstyle.
+
+Each of the functions is built upon one another, and at the last,
+exposes a single public function `extract_documentation_with_markers`
+which is used by `DocBaseClass`, to extract documentation.
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: double back ticks ``DocBaseClass``

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 3]
```diff
@@ -1,3 +1,10 @@
+"""
+Documentation comments extraction, independently of language or docstyle.
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Wrong grammar here

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 4]
```diff
@@ -1,3 +1,10 @@
...
+
+Each of the functions is built upon one another, and at the last,
+exposes a single public function `extract_documentation_with_markers`
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: public is wrong in the context of python

Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: sorry if I suggested this earlier

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 2 above)
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: also you may want to add this as a reference, see here https://stackoverflow.com/a/7754150/3578554

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 4 above)
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: also reference the function `extract_documentation_with_markers`

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 2 above)
Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Double back ticks, you mean like this ``DocBaseClass``? I haven't seen anything like this in https://stackoverflow.com/a/7754150/3578554, only :class:`Classname`

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 4 above)
Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: How can I see "compiled" documentation, taken from docstrings of panda? Should I run panda over itself? :)

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 5]
```diff
@@ -1,3 +1,10 @@
+"""
+Language and docstyle independent documentation comments extraction.
+
+Each of the functions is built upon one another, and at the last,
+exposes a single function py:function:`extract_documentation_with_markers`
```
Comment (Nosferatul) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: @Asalle you didn't add this reference https://stackoverflow.com/a/7754150/3578554

Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Sorry, I don't quite get what you mean. I didn't add a reference to function? I added py:function:`extract_documentation_with_markers`, and as far as I understood from [this](http://www.sphinx-doc.org/en/stable/markup/inline.html):
> Basically, you only need to write :role:`target`, and a link will be created to the item named target of the type indicated by role.

The so [answer](https://stackoverflow.com/questions/7753805/sphinx-automodule-how-to-reference-classes-in-same-module/7754150#7754150) does not seem to add anything but this :class: or :therm: prefix to respective entity names.

Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Sorry, I don't quite get what you mean. I didn't add a reference to function? I added py:function:`extract_documentation_with_markers`, and as far as I understood from [this](http://www.sphinx-doc.org/en/stable/markup/inline.html):
> Basically, you only need to write :role:`target`, and a link will be created to the item named target of the type indicated by role.

The so [answer](https://stackoverflow.com/questions/7753805/sphinx-automodule-how-to-reference-classes-in-same-module/7754150#7754150) does not seem to add anything but this :class: or :therm: prefix to respective entity names.

Also, when I run sphinx over coala, I see only functions docs in DocumentationExtraction module, but nothing file-level, is it okay?

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 6]
```diff
@@ -1,3 +1,10 @@
+"""
+Language and docstyle independent documentation comments extraction.
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: this still doesn't sound right, how about

`.... independent extraction of documenation comments.`

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 5 above)
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: > Sorry, I don't quite get what you mean. I didn't add a reference to function? I added py:function:extract_documentation_with_markers

Sorry this got a little function, you dont need the actual `py` prefix, also the shorthand is `func`, so something like 
```
:func:`extract_documentation.......`
``` 
should work. Also class reference:

```
:class:`.DocBaseClass`
```

Notice the dot before the class name. that helps sphinx search the class in the same module :)

Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: > Also, when I run sphinx over coala, I see only functions docs in DocumentationExtraction module, but nothing file-level, is it okay?

It should be there :) What are you using to build the docs?

Comment (Asalle) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Thanks for the clarification!
I use sphinx-apidoc and sphinx-build

Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: you do not need to do it manually, coala has helper commands :)

Just do `pip install -r docs-requirements.txt` then `./setup.py docs`, that should build the docs

**Comment by SanketDG** [comment]
ack e20bc9c

**Comment by SanketDG** [comment]
needs rebase

**Comment by Makman2** [comment]
unack e20bc9c

**Comment by Makman2** [comment]
Please use the filename you modified as a tag in the commit message. Also use `Closes XXXX` instead of `Fixes XXXX`, as the referenced issue is not marked as `bug`.

**Comment by Makman2** [comment]
If the commit message gets too long, you can strip off the part with "... describing functions", as this is usually always intended with docstrings :)

**Comment by SanketDG** [comment]
@Asalle please see this https://asciinema.org/a/104829 on how to remove the specific commit before your commit, and then rebase https://api.coala.io/en/latest/Developers/Git_Basics.html#rebasing


if you need help, you can always ask on the chat!

**Comment by Asalle** [comment]
@SanketDG thanks for the tip :)
What commit do you want me to remove? I only see my single commit in log (the one with documentation).

Rebased :)

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 7]
```diff
@@ -1,3 +1,10 @@
+"""
+Language and docstyle independent extraction of documenation comments.
+
+Each of the functions is built upon one another, and at the last,
+exposes a single function :func:`extract_documentation_with_markers`
+which is used by :class:`.DocBaseClass`, to extract documentation.
+"""
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: newline after this according to pep257 :smile:

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/4624/commits/fcf9ef3f552ecb8d3eb6a20669fdbb914c6e8c9b

**Comment by sils** [comment]
@gitmate-bot rebase

**Comment by sils** [comment]
@rultor merge

**Comment by sils** [comment]
Hey, this is your first contribution, right? Congrats and welcome to the coalaians! :) You've done your first step to become a coala developer.

Rultor will now automatically release your changes as a prerelease so users can experience the goodness of your changes right now.

We would really appreciate it if you could help us with

- [telling us about your experience](http://coala.io/newform)
- [some code reviews](http://coala.io/reviewing) for other newcomers
- fixing at least [a difficulty/low issue](http://coala.io/low)
- fill our [usability survey](http://coala.io/usability)

If you've done those and tell us about it so we can promote you to a coala developer!

We're all volunteers and we're currently struggling to keep this up. Helping all those newcomers is a lot of work and we do need your help so we can continue this for others!

**Pull request merged**